### PR TITLE
Introduce classes configuration to make sku property configurable per bucket

### PR DESCRIPTION
**Pull request by BeTheFlow95**
This PR adds the possibility to change the SKU type for each object-storage individually. Similar to setting the accessibility of object-storages

### CODE REVIEW

**Review by lieberlois — APPROVED** [review]
LGTM!

**Review by swoehrl-mw — CHANGES_REQUESTED** [review]
SKU is something Azure-specific and I want to avoid having anything provider-specific in the CRDs as that would break the abstraction.

Suggestion: Introduce the concept of classes similar to how the hybrid-cloud-postgresql-operator does it. These classes can then in the backend config be mapped to SKUs.

**Comment by lieberlois** [comment]
@swoehrl-mw I agree that the wording "SKU" should be changed, however i dont think "classes" really make sense. It suggests that each cloud provider has the same amount of classes, which is not the case. IMO the value of the field can be vendor specific, since the "backend" field already discriminates between cloud providers. This is the same for "backup"

**Comment by swoehrl-mw** [comment]
@lieberlois 

> It suggests that each cloud provider has the same amount of classes, which is not the case.

Not really. It puts the responsibility on the person configuring the operator and its backends to define classes that are available in all providers used in a platform or at least choose similar types.

>  since the "backend" field already discriminates between cloud providers. This is the same for "backup"

Normally the `backend` field should not be used, it is mainly there to allow choosing if several backends are available in one cluster. Nevertheless, the idea is still that users can just switch the backend or deploy to another cluster where the operator is configured with another default backend without changing any of the (other) fields. `backup` is not really specific, the operator could configure something equivalent for e.g. S3. As such I really don't want to have a field in there that the operator cannot automatically map between providers.

**Comment by lieberlois** [comment]
@swoehrl-mw I see what you mean, I just thought this could minimize complexity. Looking at similar tools like Crossplane, the cleanest solution would probably be to add a CRD like "ProviderConfig" or so, that maps predefined classes to vendor-specific SKUs etc. For simplicity purposes, I would be fine with adding the mapping to the config aswell though

File: hybridcloud/backends/azureblob.py [hunk 1]
```diff
@@ -490,3 +491,19 @@ def _get_user_authorized_keys(user):
...
+        return Sku(name=sku)
+
+    size_class = size_spec.get("class")
+    default_class = config_get("backends.azureblob.default_class")
+    classes = config_get("backends.azureblob.classes", default=[])
+
+    selected_class = classes[default_class]
+    if size_class and size_class in classes:
+        selected_class = classes[size_class]
+
```
Comment (swoehrl-mw) [inline hybridcloud/backends/azureblob.py:508]: To stay compatible there should be a fallback (to `"Standard_LRS"`) if neither `sku.name` nor any classes or `default_class` are defined

**Review by swoehrl-mw — CHANGES_REQUESTED** [review]

File: hybridcloud/backends/azureblob.py [hunk 2]
```diff
@@ -490,3 +491,19 @@ def _get_user_authorized_keys(user):
         ssh_public_key = SshPublicKey(description=user_key_description, key=user_public_ssh_key)
         user_authorized_keys.append(ssh_public_key)
     return user_authorized_keys
+
+
+def _determine_sku(size_spec):
+    sku = config_get("backends.azureblob.sku.name")
```
[the remark this answers is not in the document]
Comment (BeTheFlow95) [reply hybridcloud/backends/azureblob.py]: Is changed here 
81508f256296ffafdae69b681c946888afbebb50

File: hybridcloud/backends/azureblob.py [hunk 3]
```diff
@@ -490,3 +491,19 @@ def _get_user_authorized_keys(user):
...
+    if sku:
+        return Sku(name=sku)
+
+    size_class = size_spec.get("class")
+    default_class = config_get("backends.azureblob.default_class")
+    classes = config_get("backends.azureblob.classes", default=[])
```
[the remark this answers is not in the document]
Comment (BeTheFlow95) [reply hybridcloud/backends/azureblob.py]: Is changed here 
81508f256296ffafdae69b681c946888afbebb50

File: hybridcloud/backends/azureblob.py (hunk 1 above)
Comment (BeTheFlow95) [reply hybridcloud/backends/azureblob.py]: Adjusted the whole sku determination here.
81508f256296ffafdae69b681c946888afbebb50

**Review by swoehrl-mw — APPROVED** [review]

**Pull request merged**